File: Code template/Data Structrue/0-1BFS.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;
const int MOD = 1'000'000'007;
const int inf = 0x3f3f3f3f;
const ll ll_inf = 1e18;

// 计算从 start 到各个节点的最短路长度
// 如果节点不可达，则最短路长度为 inf
// 节点编号从 0 到 n-1，边权为 0 or 1
vector<int> zero_one_bfs(int n, vector<vector<pair<int, int>>>& g, int start) {
    vector<int> dis(n, inf);
    deque<int> q;
    q.push_back(start);

    while (!q.empty()) {
        int x = q.front();
        q.pop_front();

        for (auto &[y, w] : g[x]) {
            if (dis[x] + w < dis[y]) {
                dis[y] = dis[x] + w;
                if (w == 0) {
                    q.push_front(y);
                } else {
                    q.push_back(y);
                }
            }
        }
    }
    return dis;
}

File: Code template/Data Structrue/0-1BFS.py
from collections import deque
from typing import List, Tuple
from math import inf
# 计算从 start 到各个节点的最短路长度
# 如果节点无法到达，最短路为 inf
# 节点编号从 0 到 n-1，边权为 0 or 1

def zero_one_bfs(n: int, g: List[List[Tuple[int, int]]], start: int) -> List[int]:
    dis = [inf] * n
    dis[start] = 0
    q = deque([start])

    while q:
        x = q.popleft()
        for y, w in g[x]:
            if dis[x] + w < dis[y]:
                dis[y] = dis[x] + w
                if w == 0:
                    q.appendleft(y)
                else:
                    q.append(y)
    return dis